### ISSUE TITLE
Crash on [hash] or [name] in webpack output filename

### ISSUE THREAD

**Issue by ChrisRu** [Izhaki/nodemon-webpack-plugin#6]
When you have a webpack config like: (example)
```javascript
module.exports = {
  entry: './src/index.js',
  output: {
    filename: '[name]-[hash].js',
    path: 'dist'
  }
}
```
Webpack will fail starting with this error:
```bash
npm info it worked if it ends with ok
npm info using npm@5.3.0
npm info using node@v8.4.0
npm info lifecycle app-core@1.0.0~prewatch: app-core@1.0.0
npm info lifecycle app-core@1.0.0~watch: app-core@1.0.0

> app-core@1.0.0 watch /usr/app
> NODE_ENV=development webpack --watch


Webpack is watching the files…

ts-loader: Using typescript@2.5.2 and /usr/app/tsconfig.json
/usr/app/node_modules/nodemon-webpack-plugin/dist/index.js:14
    var absoluteFileName = asset.existsAt;
                                ^

TypeError: Cannot read property 'existsAt' of undefined
    at getOutputFileMeta (/usr/app/node_modules/nodemon-webpack-plugin/dist/index.js:14:33)
    at Compiler.<anonymous> (/usr/app/node_modules/nodemon-webpack-plugin/dist/index.js:43:46)
    at Compiler.applyPluginsAsyncSeries1 (/usr/app/node_modules/tapable/lib/Tapable.js:222:13)
    at Compiler.afterEmit (/usr/app/node_modules/webpack/lib/Compiler.js:365:9)
    at require.forEach.err (/usr/app/node_modules/webpack/lib/Compiler.js:354:15)
    at /usr/app/node_modules/async/dist/async.js:421:16
    at iteratorCallback (/usr/app/node_modules/async/dist/async.js:998:13)
    at /usr/app/node_modules/async/dist/async.js:906:16
    at /usr/app/node_modules/graceful-fs/graceful-fs.js:43:10
    at FSReqWrap.oncomplete (fs.js:135:15)
npm info lifecycle app-core@1.0.0~watch: Failed to exec watch script
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! app-core@1.0.0 watch: `NODE_ENV=development webpack --watch`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the app-core@1.0.0 watch script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```
`getOutputFileMeta` will give an error, because the name `[name].js` is not in the assets (only the compiled name: `index.js`).

It would be nice if the plugin could convert these filenames.

**Comment by Izhaki** [comment]
I'll have a look tonight (GMT).

**Comment by Izhaki** [comment]
Fixed in https://github.com/Izhaki/nodemon-webpack-plugin/commit/8475d3f7c3b9aecedaa2e3af61c77105d94b32c0

**Comment by Mykybo** [comment]
I was using express with hot middleware, which bundled webpack config along with webpack into the built file. After starting the file with node builtFile.js you get the error. 

I just removed the middleware since I don't need it, now it works as expected.

My webpack config is minimal, your plugin used with no config.

The hot middleware worked when I removed your plugin.

Not at pc anymore, but it should be easy to reproduce.

**Comment by Anupheaus** [comment]
Hmm, I'm finding that when I have a context set in my webpack config then nodemon-webpack-plugin works, but if I take it out then it doesn't work.

FYI, adding the console.log line like you suggested yields this response:
`{}`
Not sure why it is an empty object?

**Comment by Izhaki** [comment]
@Mykybo @Anupheaus Any chance you share a repo so I can reproduce?

**Comment by rgbyford** [comment]
I hacked the file as Izhaki requested.  The console.log produces {}

**Comment by rgbyford** [comment]
@Izhaki - sorry.  Forgot @.

**Comment by Izhaki** [comment]
That will explain the error but no idea why there are no assets...

Can I see the webpack config?

**Comment by rgbyford** [comment]
I warn you.  I am no expert!

const fs = require('fs');
const path = require('path');
const NodemonPlugin = require('nodemon-webpack-plugin');

const nodeModules = {};
fs.readdirSync('node_modules')
  .filter(function(x) {
    return ['.bin'].indexOf(x) === -1;
  })
  .forEach(function(mod) {
    nodeModules[mod] = 'commonjs ' + mod;
  });

module.exports = {
  mode: 'development',
  entry: './server/server.ts',
  output: {
    path: path.join(__dirname, '/build'),
    filename: 'bundle.js'
  },
  resolve: {
    extensions: ['.webpack.js', '.web.js', '.ts', '.tsx', '.js']
  },
  module: {
    rules: [{
      test: /\.tsx?$/, // include .js files
      enforce: "pre", // preload the jshint loader
      exclude: /node_modules/, // exclude any and all files in the node_modules folder
      use: [{
        loader: 'awesome-typescript-loader'
        // more options in the optional jshint object
        //options: {  // ⬅ formally jshint property
        //  camelcase: true,
        //  emitErrors: false,
        //  failOnHint: false
        //}
      }]
    }]
  },
  target: 'node',
  // the webpack config just works
node: {
  __dirname: false,
  __filename: false,
},
  externals: nodeModules,
  plugins: [
    new NodemonPlugin()
  ]
};

**Comment by Izhaki** [comment]
OK. That is useful - can't really see any obvious issues with this webpeck config.

Would anyone kindly share the whole compilation log? That means changing `console.log(compilation.assets)` > `console.log(compilation)`.

So hack `/node_modules/nodemon-webpack-plugin/dist/webpack-utils.js` and have a console.log like so:

```js
var getOutputFileMeta = function getOutputFileMeta(compilation) {
  console.log(compilation);
  var outputFileName = getOutputFileName(compilation);
  ...
};
```

**Comment by rgbyford** [comment]
[compilation.txt](https://github.com/Izhaki/nodemon-webpack-plugin/files/3193580/compilation.txt)

You asked.  :-)

**Comment by Izhaki** [comment]
That's odd, because in compilation you attach `assets` seem not to be empty:

```
  assets:
   { 'bundle.js': ...
```

Are you sure you got the issue mentioned (`Cannot read property 'existsAt' of undefined`) with this compilation?

**Comment by rgbyford** [comment]
I was going to say yes, because I'm pretty sure I saw the error with the "run" that I logged the compilation on.  But I made some code changes and the error's gone away!  More if it comes back, I guess.

**Comment by rgbyford** [comment]
Well, it's back.  :-(  Occurred when I tried to deploy my app to my server.  It complained that nodemon-webpack-plugin was missing.  Installed it -g.  No difference.  Installed it locally, and the complaint went away but the existsAt error is back.  I should be able to modify the plugin to log compilation.

**Comment by rgbyford** [comment]
[compilation2.txt](https://github.com/Izhaki/nodemon-webpack-plugin/files/3204479/compilation2.txt)

**Comment by rgbyford** [comment]
I don't know what's going on.  It must be something to do with a screw-up in my deploying.  Ignore me.

**Comment by sarink** [comment]
I had this happen when my entry file was incorrect, double check it!

**Comment by iKonrad** [comment]
Same here.

My config:

```
const path = require('path');
const webpack = require('webpack');
const webpackNodeExternals = require('webpack-node-externals');
const NodemonPlugin = require('nodemon-webpack-plugin');
const PATHS = require('./settings/paths.js');

const serverConfig = {
  // Inform webpack that we're building a bundle
  // for Node JS rather than for the browser
  target: 'node',
  node: {
    __dirname: false,
    __filename: false,
  },

  resolve: {
    alias: PATHS,
    extensions: ['.ts', '.tsx', '.js', '.jsx'],
  },

  // Tell webpack if we're on development or production environment
  mode: 'development',

  // Tell webpack the root file of our server application
  entry: './src/index.ts',

  // Tell webpack where to put the output file that is generated
  output: {
    filename: 'bundle.js',
    path: path.resolve(__dirname, 'build'),
  },

  // Tell webpack to run babel on every file it runs through
  module: {
    rules: [
      {
        test: /\.tsx?$/,
        exclude: /node_modules/,
        use: [{ loader: 'awesome-typescript-loader' }],
      },
      {
        test: /\.js?$/,
        loader: 'babel-loader',
        exclude: /node_modules/,
        options: {
          babelrc: false,
          presets: [
            ['env',
              {
                modules: false,
                targets: { browsers: ['last 2 versions'] },
              },
            ],
            'react',
            'stage-0',
          ],
          plugins: [
            'transform-decorators-legacy',
            'transform-class-properties',
            'syntax-class-properties',
          ],
        },
      },
    ],
  },

  plugins: [
    new NodemonPlugin(),
    new webpack.DefinePlugin({
      'process.env': {
        // Optimise React, etc
        NODE_ENV: JSON.stringify('development'),
        DEBUG: true,
      },
    }),
  ],

  externals: [
    webpackNodeExternals(),
  ],
};

module.exports = serverConfig;

```

**Comment by iKonrad** [comment]
I just had the same issue. @rgbyford and I managed to identify the problem by looking at the `errors` object in the compilation log. In my case it was that I forgot to install `awesome-typescript-loader` but used it in the config.

I've checked your logs and found this in errors object:
`[ EntryModuleNotFoundError: Entry module not found: Error: Can't resolve './server/server.ts' in '/var/www/html/mern-ts/backend'`

are you sure there's a file `server.ts` in that location?

**Comment by Izhaki** [comment]
I'm inclined to say if something like this happens, the plugin should not be oblivious to it.

Any of you has a reproducible example?

This is really important because this type of 'mysterious' issues are exactly the type that consumes a lot of people's time.

**Comment by iKonrad** [comment]
I think you can easily reproduce this if you use some non-existent loader name, at least based on my experience

**Comment by Izhaki** [comment]
But how come the error goes amiss?

**Comment by Izhaki** [comment]
I've fixed the issue.
If there are any compilation error, the plugin just lives a short console.log and does not start nodemon.

Here is the [feature file](https://github.com/Izhaki/nodemon-webpack-plugin/blob/master/features/compilation_error.feature)

Landed as 4.1.1, which is now available on npm.

Thanks for all your time!
